dafny/lcf/def.dfy

```diff
@@ -323,253 +354,161 @@ function unify(r : Prop, g : Prop) : (res : Result<Subst>)
-      return Ok((Failure, trace));
-    }
-    if unify.port != Unify {
-      print "expected: unify\n";
-      return Err;
-    }
-
-    // 3. Applies variable assignments from unification to clauses in the rule
-    //    body and continues at #1 with the updated clauses.
-    var maybe_body := build_trace_tree(trace, level+1, bound-1);
-    if maybe_body.Err? {
-      print "recursion error\n";
-      return Err;
-    }
-    var outcome := maybe_body.val.0;
-    trace := maybe_body.val.1;
-    if outcome.Failure? {
-      continue;
+    case BuiltinOp(b, args) => {
+      var maybe_leaf := mk_leaf(BuiltinOp(b, args));
+      match maybe_leaf {
+        case Ok(thm) => {
+          print_event(head); // for debugging
+          return Ok((thm, trace'));
+        }
+        case Err => {
+          print "failed to deduce builtin\n";
+          return Err;
+        }
+      }
     }
+    case App(_, _) => { // This triggers if head's prop is a fact or a non-builtin predicate.
+      // The rule used by the prop is stored.
+      var ri: nat;
+      var maybe_ri := lookup_rule(rs, head.i);
+      match maybe_ri {
+        case Ok(index) => ri := index;
+        case Err => {
+          print "could not find rule\n";
+          return Err;
+        }
+      }
+      var r := rs[ri];
+
+      var args: seq<Thm> := []; // `args` holds the child theorems of this part of the proof tree.
+      var assignment := map[]; // `assignment` is the substitution used for holding the variable assignment at this part of the proof tree.
+      // Here, the variable information that can be determined directly from the prop is stored in `assignment`.
+      var maybe_assignment := unify(r.head, head.prop);
+      match maybe_assignment {
+        case Ok(substitution) => assignment := substitution;
+        case Err => {
+          print "could not create assignment\n";
+          return Err;
+        }
+      }
 
-    // 4. After all of the body clauses of the matched rule have either succeeded, failed, or thrown an exception:
-    //      exit is traced if all of them succeeded (meaning this rule is true).
-    //      fail is traced if any of them failed (meaning this rule is false).
-    //      exception is traced if any of them threw an exception.
-    if |trace| == 0 {
-      print "unexpected end of trace\n";
-      return Err;
-    }
-    var exit := trace[0];
-    trace := trace[1..];
-    if !exit.prop.concrete() {
-      print "non concrete exit\n";
-      return Err;
-    }
+      // The program loops through the rule's children in the search for trace events that match with them.
+      for i := |r.body| downto 0
+        invariant forall j :: 0 <= j < |trace'| ==> trace'[j].prop.concrete()
+        invariant |trace'| < |trace|
+        invariant forall j :: 0 <= j < |args| ==> args[j].wf(rs)
+      {
+        // This while loop skips over all ineligble trace events until it finds one that matches with the next child rule.
```

Comment:
How does it determine the match? I guess just by seeing if the substitutions line up?

Reply:
Yes, more or less.